src/useq/_mda_sequence.py

```diff
@@ -450,24 +468,38 @@ def iter_sequence(
                 if event_kwargs[k] is not None:  # type: ignore[literal-required]
                     event_kwargs[k] += v  # type: ignore[literal-required]
 
-        # if a position has been declared with a sub-sequence, we recurse into it
-        if position and position.sequence:
-            # determine any relative position shifts or global overrides
-            _pos, _offsets = _position_offsets(position, event_kwargs)
-            # build overrides for this position
-            pos_overrides = MDAEventDict(sequence=sequence, **_pos)  # type: ignore
-            if position.name:
-                pos_overrides["pos_name"] = position.name
-            # recurse into the sub-sequence
-            yield from iter_sequence(
-                position.sequence,
-                base_event_kwargs=event_kwargs.copy(),
-                event_kwarg_overrides=pos_overrides,
-                position_offsets=_offsets,
-            )
-            continue
+        # grab global autofocus plan (may be overridden by position-specific plan below)
+        autofocus_plan = sequence.autofocus_plan
 
-        yield MDAEvent(**event_kwargs)
+        # if a position has been declared with a sub-sequence, we recurse into it
+        if position:
+            if position.sequence:
+                # determine any relative position shifts or global overrides
+                _pos, _offsets = _position_offsets(position, event_kwargs)
+                # build overrides for this position
+                pos_overrides = MDAEventDict(sequence=sequence, **_pos)  # type: ignore
+                if position.name:
+                    pos_overrides["pos_name"] = position.name
+                # recurse into the sub-sequence
+                yield from iter_sequence(
+                    position.sequence,
+                    base_event_kwargs=event_kwargs.copy(),
+                    event_kwarg_overrides=pos_overrides,
+                    position_offsets=_offsets,
+                )
+                continue
+            # note that position.sequence may be Falsey even if not None, for example
+            # if all it has is an autofocus plan.  In that case, we don't recurse.
+            # and we don't hit the continue statement, but we can use the autofocus plan
+            elif position.sequence is not None and position.sequence.autofocus_plan:
+                autofocus_plan = position.sequence.autofocus_plan
+
+        event = MDAEvent(**event_kwargs)
+        if autofocus_plan:
+            af_event = autofocus_plan.event(event)
+            if af_event is not None:
+                yield af_event
+        yield event
```

Comment:
this is the bulk of it... the rest is just an additional indentation

Reply:
@tlambert03 that's fantastic :tada: !!!

I guess the only thing left in terms of the logic is to update the `z_pos` of the autofocus event in case of a relative `z_plan` (this way the z stage will move at the home position of the z stack) like we were doing here for example https://github.com/fdrgsp/useq-schema/blob/9584af6a21dde61da74a9405db467e47ab5c7f02/src/useq/_mda_sequence.py#L549-L550.

for example
```py
event = MDAEvent(**event_kwargs)
if autofocus_plan:
    af_event = autofocus_plan.event(event)
    if af_event is not None:
        yield af_event.copy(
            update={
                "z_pos": _get_updated_z_pos(
                    event.z_pos,
                    event.index,
                    position.sequence if position else None,
                    sequence,
                )
            }
        )
```

and add 3 simple tests like these:

```py
def test_autofocus_z_pos() -> None:
    mda = TWO_CH.replace(stage_positions=[ZPOS_200], z_plan=ZRANGE2)
    af_plan = AxesBasedAF(
        autofocus_device_name="Z", autofocus_motor_offset=50, axes=("c",)
    )
    mda = mda.replace(autofocus_plan=af_plan)
    z = [e.z_pos for e in mda]
    assert z[0] == z[4] == 200


def test_autofocus_z_pos_af_sub_sequence() -> None:
    mda = TWO_CH.replace(stage_positions=[SUB_P_AF_C], z_plan=ZRANGE2)
    z = [e.z_pos for e in mda]
    assert z[0] == z[4] == 10


def test_autofocus_z_pos_af_sub_sequence_zplan() -> None:
    mda = TWO_CH_SUBPAF_C
    z = [e.z_pos for e in mda]
    assert z[2] == z[4] == 10
```

Reply:
> `_get_updated_z_pos`

sounds good.  do you want to take a stab at that and PR against this PR? (i.e. my `merge-autofocus` branch?)

- I think the logic should go in the `AutoFocusPlan.event` method, and it should only be added to the `copy(update=` dict if necessary.
- I don't think we'll need the `sequence`/`main_sequence` anymore... it should hopefully just be at `event.sequence` by that point